contracts/staking/KeepStake.sol

```diff
@@ -3,17 +3,83 @@
+///         KEEP stake owner is easy. Resolving token grant stake owner is
+///         complicated and not possible to do on-chain from a contract external
+///         to KEEP TokenStaking contract. Keep TokenStaking knows the grant ID
+///         but does not expose it externally.
+///
+///         KeepStake contract addresses this problem by exposing
+///         operator-owner mappings snapshotted off-chain based on events and
+///         information publicly available from KEEP TokenStaking contract and
+///         KEEP TokenGrant contract. Additionally, it gives the Governance
+///         ability to add new mappings in case they are ever needed; in
+///         practice, this will be needed only if someone decides to stake their
+///         KEEP token grant in KEEP network after 2021-11-11 when the snapshot
+///         was taken.
+///
+///         Operator-owner pairs were snapshotted 2021-11-11 in the following
+///         way:
+///         1. Fetch all TokenStaking events from KEEP staking contract.
+///         2. Filter out undelegated operators.
+///         3. Filter out canceled delegations.
+///         4. Fetch grant stake information from KEEP TokenGrant for that
+///            operator to determine if we are dealing with grant delegation.
+///         5. Fetch grantee address from KEEP TokenGrant contract.
+///         6. Check if we are dealing with ManagedGrant by looking for all
+///            created ManagedGrants and comparing their address against grantee
+///            address fetched from TokenGrant contract.
+contract KeepStake is Ownable {
+    IKeepTokenStaking public immutable keepTokenStaking;
+
+    mapping(address => address) public operatorToManagedGrant;
+    mapping(address => address) public operatorToGrantee;
+
+    constructor(IKeepTokenStaking _keepTokenStaking) {
+        keepTokenStaking = _keepTokenStaking;
+    }
+
+    /// @notice Allows the Governance to set new operator-managed grant pair.
+    ///         This function should only be called for managed grants if
+    ///         the snapshot does include this pair.
+    function setManagedGrant(address operator, address managedGrant)
         external
-        view
-        returns (address)
+        onlyOwner
     {
+        operatorToManagedGrant[operator] = managedGrant;
+    }
+
+    /// @notice Allows the Governance to set new operator-grantee pair.
+    ///         This function should only be called for non-managed grants if
+    ///         the snapshot does include this pair.
+    function setGrantee(address operator, address grantee) external onlyOwner {
+        operatorToGrantee[operator] = grantee;
+    }
+
+    /// @notice Resolves KEEP stake owner for the provided operator address.
+    function resolveOwner(address operator) external view returns (address) {
+        address owner = operatorToManagedGrant[operator];
+        if (owner != address(0)) {
+            return IManagedGrant(owner).grantee();
+        }
```

Comment:
Small question: I saw that ManagedGrant is not checking owner is zero address or not in the constructor. Is that possible that resolved address will be zero and what do you think about adding check before returning for whole method?

Reply:
It is possible in theory but not in practice. We'd either need to create a grant for 0x0 which is by itself a terrible mistake in the process, or we'd need to have the grantee requesting for reassignment to 0x0 and us (grant manager) accepting this reassignment.

Reply:
Ok, if you're sure then fine

Reply:
Actually... 

I am comfortable with not checking what is returned by `ManagedGrant.grantee()` but we need to check what is returned by the final `TokenStaking.ownerOf`. In https://github.com/threshold-network/solidity-contracts/pull/27/commits/e473cb064db4131deaed786a8111d1edc59cc22e I added a check to revert if `KeepStake.resolveOwner` does not know the operator and resolves the owner to 0x0.

Without reverting, the following corner case is possible:
1. Someone stakes liquid T in T TokenStaking contract.
2. They call `TokenStaking.refreshKeepStakeOwner` for their operator.
3. Owner resolves to 0x0 because legacy staking contract does not know that operator.
4. Owner is updated to 0x0.
5. Sad face.

Ideally, we'd revert in `TokenStaking` contract but we are already concerned about its size and we need to add more functionalities there, such as disabled applications. If, after implementing disabled applications, we can move that check there, that would be ideal.

Reply:
I agree that proper place for `require` is in `TokenStaking` but free size is not enough for `disableApp` right now, so maybe after next iteration of optimizations.
Also would be great to have test that checks case `owner == address(0)`

Reply:
Oops, my fault. Please see https://github.com/threshold-network/solidity-contracts/pull/27/commits/1c346e07c79a46c3ce645458e366a29a3f6438ed.